---
title: "PancreasDE"
author: "Mike Morgan"
date: "27 October 2017"
output: html_document
---

## Differential expression analysis - pancreas islet cells

First things first is to load all of the necessary packages.  To properly recapitulate the manuscript figure for the DE analysis, I'll add the ggplot themes and functions.  I'll also define a couple of convenience functions for later.

```{r, echo=TRUE, warning=FALSE, message=FALSE}
# load required packages for generating manuscript figures
library(ggplot2)
library(limma)
library(RColorBrewer)
library(goseq)
library(org.Hs.eg.db)
library(flashClust)
library(VennDiagram)
library(biomaRt)
library(stringr)
library(WGCNA)
library(ggrepel)
library(cowplot)

# setup ggplot theme
theme_mike <- function(base_size=14) {
  library(grid)
  library(ggthemes)
  (theme_foundation(base_size=base_size)
  + theme(plot.title = element_text(face = "bold",
                                    size = rel(1.2), hjust = 0.5),
          text = element_text(size=16),
          panel.background = element_rect(colour = NA),
          plot.background = element_rect(colour = NA),
          panel.border = element_rect(colour = NA),
          axis.title = element_text(face = "bold",size = rel(1)),
          axis.title.y = element_text(angle=90,vjust =2),
          axis.title.x = element_text(vjust = -0.2),
          axis.text = element_text(), 
          axis.line = element_line(colour="black"),
          axis.ticks = element_line(),
          panel.grid.major = element_line(colour="#f0f0f0"),
          panel.grid.minor = element_blank(),
          legend.key = element_rect(colour = NA),
          legend.position = "bottom",
          legend.direction = "horizontal",
          legend.key.width= unit(1.0, "cm"),
          legend.key.height = unit(0.5, "cm"),
          legend.title = element_text(face="italic"),
          plot.margin=unit(c(10,5,5,5),"mm"),
          strip.background=element_rect(colour="#f0f0f0",fill="#f0f0f0"),
          strip.text = element_text(face="bold")
  ))
}

scale_fill_Publication <- function(...){
  library(scales)
  discrete_scale("fill","Publication",
                 manual_pal(values = c("#386cb0", "#fdb462",
                                       "#7fc97f","#ef3b2c","#662506",
                                       "#a6cee3","#fb9a99","#984ea3",
                                       "#ffff33", "#b2df8a", "#3E89BE", 
                                       "#CA834E", "#518A87", "#5B113C", 
                                       "#55813B", "#E704C4", "#DDAEA2", 
                                       "#77837F", "#A53327", "#608EFF", 
                                       "#B599D7", "#A50149", "#4E0025", 
                                       "#C9B1A9",
                                       "#33a02c", "#fb9a99", "#e31a1c",
                                       "#fdbf6f", "#ff7f00", "#cab2d6",
                                       "#6a3d9a", "#ffff99", "#b15928")), ...)
  
}

scale_colour_Publication <- function(...){
  library(scales)
  discrete_scale("colour", "Publication",
                 manual_pal(values = c("#386cb0", "#fdb462",
                                       "#7fc97f","#ef3b2c","#662506",
                                       "#a6cee3","#fb9a99","#984ea3",
                                       "#ffff33", "#b2df8a", "#3E89BE", 
                                       "#CA834E", "#518A87", "#5B113C", 
                                       "#55813B", "#E704C4", "#DDAEA2", 
                                       "#77837F", "#A53327", "#608EFF", 
                                       "#B599D7", "#A50149", "#4E0025", 
                                       "#C9B1A9",
                                       "#33a02c", "#fb9a99", "#e31a1c",
                                       "#fdbf6f", "#ff7f00", "#cab2d6",
                                       "#6a3d9a", "#ffff99", "#b15928")), ...)
  
}

tsne_wrapper <- function(dataframe, perplexity=30, is.dist=FALSE){
  require(Rtsne)
  set.seed(42)
  if(is.dist){
    data.tsne <- Rtsne(t(dataframe),
                       perplexity=perplexity, is_distance=is.dist, pca=FALSE)
    sample.names <- labels(dataframe)
    
  }
  else {
    dataframe <- dataframe[, !duplicated(t(dataframe))]
    data.tsne <- Rtsne(t(dataframe), perplexity=perplexity)
    sample.names <- colnames(dataframe)
  }
  data.map <- data.frame(data.tsne$Y)  
  colnames(data.map) <- c("Dim1", "Dim2")
  data.map$Sample <- sample.names
  
  return(data.map)
}
```

The raw data should have been downloaded with `DownloadData.sh`, and the batch corrected data should have been generated by running `source("Pancreas/preparedata.R")` followed by `source("Pancreas/FourPlots_panc.R")` in an R session.  If you are running this notebook in an active RStudio session you can just source these two R scripts before executing the rest of this notebook.

First we'll load all of the raw and batch corrected data into the R environment.  This might take a little while as these are quite large data sets (1000s of cells each).

```{r, echo=TRUE, warning=FALSE, message=FALSE}
# all paths assume code is executed in the MNN2017 repository directory
# change these as necessary
# all corrected data are in a single file
all.correct <- read.table("../Pancreas/Data/mnnCorrected.tsv",
                          h=TRUE, sep="\t", stringsAsFactors=FALSE)
rownames(all.correct) <- all.correct$gene_id

gse81076.raw <- read.table("../Pancreas/Data/GSE81076_SFnorm.tsv", sep="\t",
                           h=T, stringsAsFactors=F)

gse85241.raw <- read.table("../Pancreas/Data/GSE85241_SFnorm.tsv", sep="\t",
                           h=T, stringsAsFactors=F)

emtab5061.raw <- read.table("../Pancreas/Data/E-MTAB-5061_SFnorm.tsv", sep="\t",
                           h=T, stringsAsFactors=F)

gse86473.raw <- read.table("../Pancreas/Data/GSE86473_SFnorm.tsv", sep="\t",
                               h=T, stringsAsFactors=F)

all.genes <- unique(all.correct$gene_id)

ensembl <- useEnsembl(biomart='ENSEMBL_MART_ENSEMBL', dataset='hsapiens_gene_ensembl', GRCh=37)

gene_symbol <- getBM(attributes=c('ensembl_gene_id', 'external_gene_name'),
                     filters='external_gene_name', mart=ensembl,
                     values=all.genes)
```

We also need to retrieve the meta data and highly variable genes, which are downloaded from the FTP server along with the highly variable gene sets and the raw data.

```{r, echo=TRUE, warning=FALSE, message=FALSE}
# meta data
gse81076.meta <- read.table("../Pancreas/Data/GSE81076_marker_metadata.tsv", sep="\t",
                            stringsAsFactors=F, h=TRUE)

gse85241.meta <- read.table("../Pancreas/Data/GSE85241_marker_metadata.tsv", sep="\t",
                            stringsAsFactors=F, h=T)

emtab5061.meta <- read.table("../Pancreas/Data/E-MTAB-5061_metadata.tsv", sep="\t",
                            stringsAsFactors=F, h=T)
emtab5061.meta$Study <- "EMTAB5061"

gse86473.meta <- read.table("../Pancreas/Data/GSE86473_metadata.tsv", sep="\t",
                            stringsAsFactors=F, h=T)

gse86473.hvg <- read.table("../Pancreas/Data/GSE86473-HVG.tsv", h=F, sep="\t",
                           stringsAsFactors=F)

gse81076.hvg <- read.table("../Pancreas/Data/GSE81076-HVG.tsv", h=F, sep="\t",
                           stringsAsFactors=F)

emtab5061.hvg <- read.table("../Pancreas/Data/E-MTAB-5061-HVG.tsv", h=F, sep="\t",
                           stringsAsFactors=F)

gse85241.hvg <- read.table("../Pancreas/Data/GSE85241-HVG.tsv", h=F, sep="\t",
                           stringsAsFactors=F)
```


The comparison of the two smallest cell populations would be in gse81076, between the delta islets and the PP cells.  We want to naively cluster the cells together and find the cells that relate to say, PP and delta cells based on marker genes, etc.  This would essentially reflect the sort of work-flow that would be performed when the cell type labels are not known _a priori_.  Then using the same cells on the uncorrected data, we want to show that we can find more DE genes after correction, clustering, etc.

```{r, echo=TRUE, warning=FALSE, message=FALSE}
merged.meta <- data.frame(do.call(rbind, list("GSE81076"=gse81076.meta[, c("Sample", "Study", "CellType")],
                                              "GSE85241"=gse85241.meta[, c("Sample", "Study", "CellType")],
                                              "EMTAB5061"=emtab5061.meta[, c("Sample", "Study", "CellType")],
                                              "GSE86473"=gse86473.meta[, c("Sample", "Study", "CellType")])))
merged.meta$CellType <- gsub(tolower(merged.meta$CellType), pattern=" cell", replacement="")

table(merged.meta$CellType)
```

## Cell clustering
We now want to cluster cells and try to find their cell label(s), before doing a DE analysis on those cells (using the original uncorrected expression values).

### Clustering using MNN corrected data 

```{r, echo=TRUE, warning=FALSE, message=FALSE}
# remove the small scrappy cell types, there are at least 30 cells
inc.types <- c("pp", "alpha", "beta", "acinar", "delta", "ductal", "mesenchyme", "psc", "gamma")
inc.cells <- merged.meta$Sample[merged.meta$CellType %in% inc.types]

# cluster cells on the hvg
all.cells <- all.correct[, intersect(inc.cells, colnames(all.correct))]
all.cells[is.na(all.cells)] <- 0
all.cells <- all.cells[, intersect(colnames(all.cells), merged.meta$Sample)]
merged.meta <- merged.meta[merged.meta$Sample %in% intersect(colnames(all.cells), merged.meta$Sample), ]
rownames(merged.meta) <- merged.meta$Sample

# get the cell IDs for each study
gse81076.cells <- merged.meta$Sample[merged.meta$Study == "GSE81076"]
gse86473.cells <- merged.meta$Sample[merged.meta$Study == "GSE86473"]
emtab5061.cells <- merged.meta$Sample[merged.meta$Study == "EMTAB5061"]
gse85241.cells <- merged.meta$Sample[merged.meta$Study == "GSE85241"]
```


```{r, echo=TRUE, warning=FALSE, message=FALSE}
# NB: this is a slow operation as there are ~7000 cells to embed
set.seed(42)
all.tsne <- tsne_wrapper(all.cells, perplexity=50)
all.tsne$Sample <- colnames(all.cells)
all.merge <- merge(all.tsne, merged.meta, by='Sample')
```


```{r, echo=TRUE, warning=FALSE, message=FALSE}
# try a k-means with k=8 as we expect ~8 different cell types across all 4 data sets
set.seed(42)
all.kmeans <- kmeans(all.tsne[, 1:2], centers=8)
kmean.df <- data.frame(cbind(colnames(all.cells), all.kmeans$cluster))
colnames(kmean.df) <- c("Sample", "Kmeans")

kmean.merge <- merge(all.merge, kmean.df, by='Sample')
```

Let's plot the tSNE, colour by the kmeans partition.  Please note that since the original manuscript submission minor alterations in the MNNCorrect code have resulted in qualitatively different plots.  The same partitioning is possible, but due to the stochastic nature of tSNE, small changes in values can propagate into apparently large differences in tSNE visualisation.

```{r, echo=TRUE, warning=FALSE, message=FALSE, fig.height=4.5, fig.width=9.5}
by.cell <- ggplot(kmean.merge, aes(x=Dim1, Dim2, colour=CellType)) +
  geom_point(size=1, alpha=0.7) + theme_mike() +
  scale_colour_Publication() +
  labs(x="tSNE 1", y="tSNE 2")

by.study <- ggplot(kmean.merge, aes(x=Dim1, Dim2, colour=Kmeans)) +
  geom_point(size=1, alpha=0.6) + theme_mike() +
  scale_colour_Publication() +
  labs(x="tSNE 1", y="tSNE 2")

plot_grid(by.cell, by.study)
```

K-means on the tSNE dimensions does pretty well, it splits some of the larger clusters though, but it is obvious these should be merged into a single cluster.  Let's overlay the different marker genes from the original publications onto these cells to identify what types they are.

```{r, echo=TRUE, message=FALSE, warning=FALSE, fig.width=9, fig.height=9}
marker.genes <- c("SST", "PPY", "PRSS1", "KRT19", "INS", "GHRL", "GCG", "ESAM", "COL1A1")
marker.exprs <- data.frame(t(all.cells[marker.genes, ]))
marker.exprs$Sample <- rownames(marker.exprs)

meta.marker <- merge(kmean.merge, marker.exprs, by='Sample')

sst.p <- ggplot(meta.marker, aes(x=Dim1, Dim2, colour=SST)) +
  geom_point(size=1, alpha=0.4) + theme_mike() +
  scale_colour_distiller(palette="RdYlBu", direction=-1) +
  guides(shape=FALSE)

ppy.p <- ggplot(meta.marker, aes(x=Dim1, Dim2, colour=PPY)) +
  geom_point(size=1, alpha=0.4) + theme_mike() +
  scale_colour_distiller(palette="RdYlBu", direction=-1) +
  guides(shape=FALSE)

prss1.p <- ggplot(meta.marker, aes(x=Dim1, Dim2, colour=PRSS1)) +
  geom_point(size=1, alpha=0.4) + theme_mike() +
  scale_colour_distiller(palette="RdYlBu", direction=-1) +
  guides(shape=FALSE)

krt19.p <- ggplot(meta.marker, aes(x=Dim1, Dim2, colour=KRT19)) +
  geom_point(size=1, alpha=0.4) + theme_mike() +
  scale_colour_distiller(palette="RdYlBu", direction=-1) +
  guides(shape=FALSE)

ins.p <- ggplot(meta.marker, aes(x=Dim1, Dim2, colour=INS)) +
  geom_point(size=1, alpha=0.4) + theme_mike() +
  scale_colour_distiller(palette="RdYlBu", direction=-1) +
  guides(shape=FALSE)

gcg.p <- ggplot(meta.marker, aes(x=Dim1, Dim2, colour=GCG)) +
  geom_point(size=1, alpha=0.4) + theme_mike() +
  scale_colour_distiller(palette="RdYlBu", direction=-1) +
  guides(shape=FALSE)

esam.p <- ggplot(meta.marker, aes(x=Dim1, Dim2, colour=ESAM)) +
  geom_point(size=1, alpha=0.4) + theme_mike() +
  scale_colour_distiller(palette="RdYlBu", direction=-1) +
  guides(shape=FALSE)

col1a1.p <- ggplot(meta.marker, aes(x=Dim1, Dim2, colour=COL1A1)) +
  geom_point(size=1, alpha=0.4) + theme_mike() +
  scale_colour_distiller(palette="RdYlBu", direction=-1)

plot_grid(sst.p, ppy.p, prss1.p, krt19.p, ins.p, gcg.p, col1a1.p,
          ncol=3, labels=c("SST", "PPY", "PRSS1", "KTR19", "INS", "GCG", "COL1A1"))
```

Based on the expression of these marker genes we can surmise that:

* cluster 3 - Delta cells (SST+)
* cluster 7 - Beta cells (INS+)
* cluster 1, 4 & 5 - Alpha cells (GCG+)
* cluster 2 - Acinar cells (GCG+)
* cluster 6 - PP/Gamma cells (PPY+)
* cluster 8 - Ductal cells (KRT19+)
* cluster 3 - Mesenchyme/PSC cells (COL1A1+)

Cluster 3 are mix of delta cells and mesenchyme/PSC cells, but they are delineated by the expression of _COL1A1_ and _KRT19_ as can be seen in the plots above with the marker gene expression overlaid on the tSNE plots.  Therefore with a combination of clustering and marker genes following dataset combination and batch correction, we can re-find the original cell labels.

Let's define them explicitly:

```{r, echo=TRUE, warning=FALSE, message=FALSE}
meta.marker$new.labels <- "Unk"
meta.marker$new.labels[meta.marker$Kmeans == 3 & meta.marker$SST >= 0.1] <- "Delta"
meta.marker$new.labels[meta.marker$Kmeans == 7] <- "Beta"
meta.marker$new.labels[meta.marker$Kmeans == 1] <- "Alpha"
meta.marker$new.labels[meta.marker$Kmeans == 4] <- "Alpha"
meta.marker$new.labels[meta.marker$Kmeans == 5] <- "Alpha"
meta.marker$new.labels[meta.marker$Kmeans == 8] <- "Acinar"
meta.marker$new.labels[meta.marker$Kmeans == 6] <- "Gamma"
meta.marker$new.labels[meta.marker$Kmeans == 2] <- "Ductal"
meta.marker$new.labels[meta.marker$Kmeans == 3 & meta.marker$COL1A1 >= 0.1] <- "Mesenchyme"

table(meta.marker$CellType, meta.marker$new.labels)
print(randIndex(table(meta.marker$CellType, meta.marker$new.labels), adjust=TRUE))
```

The gamma and pp cells are the same, but different studies use different labels.  Here, as described in the manuscript, we consider them all to be of the same type, labelled `Gamma`.  We also merge together the mesenchyme and PSC cells as we cannot discriminate between them transcriptionally.  The adjusted Rand index, assuming the original cell type labels as the ground truth, is `r randIndex(table(meta.marker$CellType, meta.marker$new.labels), adjust=TRUE)`.

Below is the equivalent to Supplementary Figure4a in the manuscript.

```{r, fig.width=5, fig.height=7.5}
new.tsne <- plot_grid(by.study, by.cell, nrow=2, labels=c("a", "b"))

new.tsne
```

Now that we have re-clustered the batch correct data in a way that is agnostic to the original cell type labels, we can use these to demonstrate the improved power of combining data sets.

##  DE analysis of Delta and PP islsets - GSE81076 (CEL-Seq) - raw data

We can see the number of each cell type by each study, noting that some cell types do not appear in all studies.

```{r, echo=TRUE}
table(meta.marker$CellType, meta.marker$Study)
```

With the new labels there is still the same coverage of cell types across studies, remembering that `gamma` and `pp` cells are combined into a single category, as are `mesenchyme` and `psc` cells.

```{r, echo=TRUE}
table(meta.marker$new.labels, meta.marker$Study)
```

Firstly we use the uncorrected data from GSE81076 to compare PP/$\gamma$ and $\delta$ islet cells.  From the table above we can see that there are 15 PP/$\gamma$ and 51 $\delta$ islet cells.  We'll select just those genes with an absolute log$_2$ fold change > 1.


```{r, echo=TRUE, warning=FALSE, message=FALSE}
# select the same raw genes that are in the corrected data
# use the old cell labels from GSE81076
meta.marker$new.labels <- factor(meta.marker$new.labels,
                                labels=c("Unk", "Gamma", "Beta", "Alpha", "Acinar", "Ductal", "Mesenchyme", "Delta"),
                                levels=c("Unk", "Gamma", "Beta", "Alpha", "Acinar", "Ductal", "Mesenchyme", "Delta"))

raw.meta <- meta.marker[meta.marker$Sample %in% gse81076.cells, ]
raw.meta$CellType <- factor(raw.meta$CellType,
                            labels=c("Ductal", "Alpha", "Delta", "Beta", "Acinar", "PP", "Mesenchyme"),
                            levels=c("ductal", "alpha", "delta", "beta", "acinar", "pp", "mesenchyme"))

# set column cell ordering and meta data ordering the same
rownames(raw.meta) <- raw.meta$Sample

raw.design <- model.matrix(~ 0 + CellType, data=raw.meta)
# raw.design <- model.matrix(~ 0 + new.labels, data=raw.meta)

rownames(gse81076.raw) <- gse81076.raw$gene_id
raw.exprs <- gse81076.raw[gse81076.raw$gene_id %in% all.genes, ]
raw.exprs <- raw.exprs[, colnames(gse81076.raw) %in% gse81076.cells]
raw.exprs <- raw.exprs[,colnames(raw.exprs)!="gene_id"]
raw.exprs2 <- raw.exprs[, rownames(raw.meta)]

colnames(raw.design) <- levels(raw.meta$CellType)

# fit a linear model using limma
raw.fit <- limma::lmFit(as.matrix(raw.exprs2),
                        raw.design)

contrast.mat <-  limma::makeContrasts(contrasts="Delta-PP",
                                      levels=raw.design)

raw.fit <- limma::contrasts.fit(raw.fit, contrasts=contrast.mat)

# just select genes with a log Fold change >= 1.0
raw.fit <- limma::treat(raw.fit, lfc=1)
raw.sum.res <- summary(limma::decideTests(raw.fit))

knitr::kable(raw.sum.res)
```

There are just 14 genes differentially expressed between $\delta$ and PP/$\gamma$ islet cells.  Let's check the identity of these 14 genes; they should at the very least include the two marker genes _SST_ and _PPY_.

```{r, echo=TRUE, warning=FALSE, message=FALSE}
raw.de.res <- limma::topTable(raw.fit, n=Inf, sort="p", p=1.0)
raw.de.res$Sig <- 0
raw.de.res$Sig[raw.de.res$adj.P.Val <= 0.05] <- 1
raw.de.res$Sig <- as.factor(raw.de.res$Sig)

raw.de.res$Diff <- 0
raw.de.res$Diff[raw.de.res$logFC < 0 & raw.de.res$Sig == 1] <- -1
raw.de.res$Diff[raw.de.res$logFC > 0 & raw.de.res$Sig == 1] <- 1

raw.de.res$gene_id <- rownames(raw.de.res)
raw.de.merge <- merge(raw.de.res,  gene_symbol,
                  by.x='gene_id', by.y='external_gene_name', all.x=TRUE)
raw.gene.top <- raw.de.merge$gene_id
raw.gene.top[raw.de.merge$logFC > -3 & raw.de.merge$logFC < 3] <- ""
raw.gene.top[is.na(raw.gene.top)] <- ""

ma <- ggplot(raw.de.merge, aes(x=AveExpr, y=logFC, colour=Sig)) +
  geom_point(size=1, alpha=0.8) + theme_mike() +
  scale_colour_Publication() +
  labs(x=expression(paste("Mean log"[2], " Expression")),
       y=expression(paste("log"[2], " Fold Change")),
       title="GSE81076 Delta vs. PP cells") +
  guides('colour'=guide_legend(title="Statistically significant")) +
  geom_text_repel(aes(label=raw.gene.top),
                  colour='black')

ma
```

```{r, echo=TRUE, message=FALSE, warning=FALSE}
unique(raw.de.merge$gene_id[raw.de.merge$Sig == 1])
```

As expected the two most DE genes are _PPY_ (marker gene for PP cells) and _SST_ (marker gene for delta cells).  There is a TF (_PAX6_) that relevant which is down-regulated in the delta cells relative to the PP cells (it is a master TF for PP cells).  Apparently _PAX6_ is required for the maintenance of islet cell function [http://journals.plos.org/plosone/article?id=10.1371/journal.pone.0054173], and it is a determinant of islet cell development from their Ngn3+ endocrine progenitor cells.

Now let's look at what happens when we use the corrected data for all 4 data sets.

### GSE85241 DE analysis

```{r, echo=FALSE, warning=FALSE, message=FALSE}
# select the same raw genes that are in the corrected data
# use the newly defined cell labels

meta.marker$new.labels <- factor(meta.marker$new.labels,
                                labels=c("Unk", "Gamma", "Beta", "Alpha", "Acinar", "Ductal", "Mesenchyme", "Delta"),
                                levels=c("Unk", "Gamma", "Beta", "Alpha", "Acinar", "Ductal", "Mesenchyme", "Delta"))

raw.meta <- meta.marker[meta.marker$Sample %in% gse85241.cells, ]
raw.meta$CellType <- factor(raw.meta$CellType,
                            labels=c("Ductal", "Alpha", "Delta", "Beta", "Acinar", "PP", "Mesenchyme"),
                            levels=c("ductal", "alpha", "delta", "beta", "acinar", "pp", "mesenchyme"))

# set column cell ordering and meta data ordering the same
rownames(raw.meta) <- raw.meta$Sample

raw.design <- model.matrix(~ 0 + CellType, data=raw.meta)
# raw.design <- model.matrix(~ 0 + new.labels, data=raw.meta)

rownames(gse85241.raw) <- gse85241.raw$gene_id
raw.exprs <- gse85241.raw[gse85241.raw$gene_id %in% all.genes, ]
raw.exprs <- raw.exprs[, colnames(gse85241.raw) %in% gse85241.cells]
raw.exprs <- raw.exprs[,colnames(raw.exprs)!="gene_id"]
raw.exprs2 <- raw.exprs[, rownames(raw.meta)]

colnames(raw.design) <- levels(raw.meta$CellType)
# colnames(raw.design) <- c("Unk", "PP", "Beta", "Alpha", "Acinar", "Ductal", "Mesenchyme", "Delta")

# fit a linear model
raw.fit <- limma::lmFit(as.matrix(raw.exprs2),
                        raw.design)


contrast.mat <-  limma::makeContrasts(contrasts="Delta-PP",
                                      levels=raw.design)

raw.fit <- limma::contrasts.fit(raw.fit, contrasts=contrast.mat)
raw.fit <- limma::treat(raw.fit, lfc=1)
raw.sum.res <- summary(limma::decideTests(raw.fit))

knitr::kable(raw.sum.res)
```

There are 42 genes DE in the GSE85241 study.

```{r, echo=FALSE, warning=FALSE, message=FALSE}
gse85241.raw.de.res <- limma::topTable(raw.fit, n=Inf, sort="p", p=1.0)
gse85241.raw.de.res$Sig <- 0
gse85241.raw.de.res$Sig[gse85241.raw.de.res$adj.P.Val <= 0.05] <- 1
gse85241.raw.de.res$Sig <- as.factor(gse85241.raw.de.res$Sig)

gse85241.raw.de.res$Diff <- 0
gse85241.raw.de.res$Diff[gse85241.raw.de.res$logFC < 0 & gse85241.raw.de.res$Sig == 1] <- -1
gse85241.raw.de.res$Diff[gse85241.raw.de.res$logFC > 0 & gse85241.raw.de.res$Sig == 1] <- 1

gse85241.raw.de.res$gene_id <- rownames(gse85241.raw.de.res)
gse85241.raw.de.merge <- merge(gse85241.raw.de.res,  gene_symbol,
                  by.x='gene_id', by.y='external_gene_name', all.x=TRUE)
gse85241.raw.gene.top <- gse85241.raw.de.merge$gene_id
gse85241.raw.gene.top[gse85241.raw.de.merge$logFC > -5 & gse85241.raw.de.merge$logFC < 5] <- ""
gse85241.raw.gene.top[is.na(gse85241.raw.gene.top)] <- ""

gse85241.ma <- ggplot(gse85241.raw.de.merge, aes(x=AveExpr, y=logFC, colour=Sig)) +
  geom_point(size=1, alpha=0.8) + theme_mike() +
  scale_colour_Publication() +
  labs(x=expression(paste("Mean log"[2], " Expression")),
       y=expression(paste("log"[2], " Fold Change")),
       title="GSE85241 Delta vs. PP cells") +
  guides('colour'=guide_legend(title="Statistically significant")) +
  geom_text_repel(aes(label=gse85241.raw.gene.top),
                  colour='black')

gse85241.ma
```

### GSE86473 DE analysis

```{r, echo=FALSE, warning=FALSE, message=FALSE}
# select the same raw genes that are in the corrected data
# use the newly defined cell labels

meta.marker$new.labels <- factor(meta.marker$new.labels,
                                labels=c("Unk", "Gamma", "Beta", "Alpha", "Acinar", "Ductal", "Mesenchyme", "Delta"),
                                levels=c("Unk", "Gamma", "Beta", "Alpha", "Acinar", "Ductal", "Mesenchyme", "Delta"))

raw.meta <- meta.marker[meta.marker$Sample %in% gse86473.cells, ]
raw.meta$CellType <- factor(raw.meta$CellType,
                            labels=c("Ductal", "Alpha", "Delta", "Beta", "Acinar", "PP", "Mesenchyme"),
                            levels=c("ductal", "alpha", "delta", "beta", "acinar", "pp", "mesenchyme"))

# set column cell ordering and meta data ordering the same
rownames(raw.meta) <- raw.meta$Sample

raw.design <- model.matrix(~ 0 + CellType, data=raw.meta)
# raw.design <- model.matrix(~ 0 + new.labels, data=raw.meta)

rownames(gse86473.raw) <- gse86473.raw$gene_id
raw.exprs <- gse86473.raw[gse86473.raw$gene_id %in% all.genes, ]
raw.exprs <- raw.exprs[, colnames(gse86473.raw) %in% gse86473.cells]
raw.exprs <- raw.exprs[,colnames(raw.exprs)!="gene_id"]
raw.exprs2 <- raw.exprs[, rownames(raw.meta)]

colnames(raw.design) <- levels(raw.meta$CellType)
# colnames(raw.design) <- c("Unk", "PP", "Beta", "Alpha", "Acinar", "Ductal", "Mesenchyme", "Delta")

# fit a linear model
raw.fit <- limma::lmFit(as.matrix(raw.exprs2),
                        raw.design)


contrast.mat <-  limma::makeContrasts(contrasts="Delta-PP",
                                      levels=raw.design)

raw.fit <- limma::contrasts.fit(raw.fit, contrasts=contrast.mat)
raw.fit <- limma::treat(raw.fit, lfc=1)
raw.sum.res <- summary(limma::decideTests(raw.fit))

knitr::kable(raw.sum.res)
```

There are 59 DE genes in the GSE86473 study between $\delta$-islets and PP cells.

```{r, echo=FALSE, warning=FALSE, message=FALSE}
gse86473.raw.de.res <- limma::topTable(raw.fit, n=Inf, sort="p", p=1.0)
gse86473.raw.de.res$Sig <- 0
gse86473.raw.de.res$Sig[gse86473.raw.de.res$adj.P.Val <= 0.05] <- 1
gse86473.raw.de.res$Sig <- as.factor(gse86473.raw.de.res$Sig)

gse86473.raw.de.res$Diff <- 0
gse86473.raw.de.res$Diff[gse86473.raw.de.res$logFC < 0 & gse86473.raw.de.res$Sig == 1] <- -1
gse86473.raw.de.res$Diff[gse86473.raw.de.res$logFC > 0 & gse86473.raw.de.res$Sig == 1] <- 1

gse86473.raw.de.res$gene_id <- rownames(gse86473.raw.de.res)
gse86473.raw.de.merge <- merge(gse86473.raw.de.res,  gene_symbol,
                  by.x='gene_id', by.y='external_gene_name', all.x=TRUE)
gse86473.raw.gene.top <- gse86473.raw.de.merge$gene_id
gse86473.raw.gene.top[gse86473.raw.de.merge$logFC > -5 & gse86473.raw.de.merge$logFC < 5] <- ""
gse86473.raw.gene.top[is.na(gse86473.raw.gene.top)] <- ""

gse86473.ma <- ggplot(gse86473.raw.de.merge, aes(x=AveExpr, y=logFC, colour=Sig)) +
  geom_point(size=1, alpha=0.8) + theme_mike() +
  scale_colour_Publication() +
  labs(x=expression(paste("Mean log"[2], " Expression")),
       y=expression(paste("log"[2], " Fold Change")),
       title="GSE86473 Delta vs. PP cells") +
  guides('colour'=guide_legend(title="Statistically significant")) +
  geom_text_repel(aes(label=gse86473.raw.gene.top),
                  colour='black')

gse86473.ma
```

### E-MTAB-5061 DE analysis

This data set does not contain any PP/Gamma cells.

## DE analysis on combined data - Delta vs. PP/Gamma cells

```{r, echo=TRUE}
rowSums(table(meta.marker$new.labels, meta.marker$Study))
```

The comparison is now 416 vs. 409 cells, there should be many more DE genes.

```{r, echo=FALSE, warning=FALSE, message=FALSE}
# select the same raw genes that are in the corrected data
# need to use the raw values, and block on batch

correct.genes <- rownames(all.correct)

# explicitly select the same cells that are subsampled from the corrected data
rownames(gse86473.raw) <- gse86473.raw$gene_id
gse86473.tomerge <- gse86473.raw[, colnames(gse86473.raw) %in% meta.marker$Sample]
gse86473.tomerge$gene_id <- rownames(gse86473.tomerge)
gse86473.tomerge[is.na(gse86473.tomerge)] <- 0

rownames(gse85241.raw) <- gse85241.raw$gene_id
gse85241.tomerge <- gse85241.raw[, colnames(gse85241.raw) %in% meta.marker$Sample]
gse85241.tomerge$gene_id <- rownames(gse85241.tomerge)
gse85241.tomerge[is.na(gse85241.tomerge)] <- 0

rownames(emtab5061.raw) <- emtab5061.raw$gene_id
emtab5061.tomerge <- emtab5061.raw[, colnames(emtab5061.raw) %in% meta.marker$Sample]
emtab5061.tomerge$gene_id <- rownames(emtab5061.tomerge)
emtab5061.tomerge[is.na(emtab5061.tomerge)] <- 0

rownames(gse81076.raw) <- gse81076.raw$gene_id
gse81076.tomerge <- gse81076.raw[, colnames(gse81076.raw) %in% meta.marker$Sample]
gse81076.tomerge$gene_id <- rownames(gse81076.tomerge)
gse81076.tomerge[is.na(gse81076.tomerge)] <- 0

raw.list <- list("GSE86473"=gse86473.tomerge,
                 "GSE85241"=gse85241.tomerge,
                 "EMTAB5061"=emtab5061.tomerge,
                 "GSE81076"=gse81076.tomerge)

merged.raw <- Reduce(x=raw.list,
                     f=function(x, y) merge(x, y, by='gene_id'))

rownames(merged.raw) <- merged.raw$gene_id
correct.exprs <- merged.raw
correct.exprs <- correct.exprs[, 2:(dim(correct.exprs)[2])]

# set up a new variable that captures the cell label X study interaction
meta.marker$interact.group <- as.factor(apply(meta.marker,
                                              1, FUN=function(X) paste(X["Study"], X["new.labels"], sep=".")))
rownames(meta.marker) <- meta.marker$Sample

correct.design <- model.matrix(~ 0 + interact.group, data=meta.marker)
colnames(correct.design) <- levels(meta.marker$interact.group)

# fit a linear model
correct.fit <- limma::lmFit(correct.exprs[, rownames(meta.marker)],
                            correct.design)

contrast.mat <- limma::makeContrasts(GSE81076.Delta-GSE81076.Gamma, GSE85241.Delta-GSE85241.Gamma,
                                     EMTAB5061.Delta-EMTAB5061.Gamma, GSE86473.Delta-GSE86473.Gamma,
                                     levels=correct.design)

# get the average log fold changes across batch by cell type interactions
correct.fit2 <- limma::contrasts.fit(correct.fit, contrasts=rowMeans(contrast.mat))

# just select genes with a log Fold change >= 1.0
correct.fit2 <- limma::treat(correct.fit2, lfc=1)
correct.sum.res <- summary(limma::decideTests(correct.fit2))

knitr::kable(correct.sum.res)
```

Immediately we can see that there are more genes detected as differentially expressed between PP and delta islet cells with log$_2$ fold change > 2 now; 74 in total.

```{r, echo=FALSE, warning=FALSE, message=FALSE}
# remember that the expression values are now L2 norms. Not log expression.
correct.de.res <- limma::topTable(correct.fit2, n=Inf, sort="p", p=1.0)

# set -inf to -308
correct.de.res$adj.P.Val[correct.de.res$adj.P.Val == 0] <- .Machine$double.xmin

correct.de.res$Sig <- 0
correct.de.res$Sig[correct.de.res$adj.P.Val <= 0.05] <- 1
correct.de.res$Sig <- as.factor(correct.de.res$Sig)

correct.de.res$Diff <- 0
correct.de.res$Diff[correct.de.res$logFC < 0 & correct.de.res$Sig == 1] <- -1
correct.de.res$Diff[correct.de.res$logFC > 0 & correct.de.res$Sig == 1] <- 1

correct.de.res$gene_id <- rownames(correct.de.res)
correct.de.merge <- merge(correct.de.res,  gene_symbol,
                  by.x='gene_id', by.y='external_gene_name', all.x=TRUE)
correct.gene.top <- correct.de.merge$gene_id
correct.gene.top[correct.de.merge$logFC > -3 & correct.de.merge$logFC < 3] <- ""
correct.gene.top[is.na(correct.gene.top)] <- ""

ma.correct <- ggplot(correct.de.merge, aes(x=AveExpr, y=logFC, colour=Sig)) +
  geom_point(size=1, alpha=0.8) + theme_mike() +
  scale_colour_Publication() +
  labs(x=expression(paste("Mean log"[2], "Normalized Expression")),
       y=expression(paste("log"[2], " Fold Change")),
       title="MNN corrected PP vs. Delta cells") +
  guides('colour'=guide_legend(title="Statistically significant")) +
  geom_text_repel(aes(label=correct.gene.top),
                  colour='black')

ma.correct
```

We still see _SST_ and _PPY_ as the principal differences between delta and PP cells, which indicates nothing stupid is happening.  Additionally there are genes such as _GC_ (vitamin D binding protein) and _LEPR_ leptin receptor that are now DE.  In fact, we can look at how the two sets of DE genes intersect; the raw data DE genes should be a subset of the MNN corrected genes.

Let's also have a volcano plot for both raw and corrected data.

```{r, echo=FALSE, warning=FALSE, message=FALSE, fig.height=4.5, fig.width=7.5}
raw.volcano <- ggplot(raw.de.merge, aes(x=logFC, y=-log10(adj.P.Val))) +
  geom_point(alpha=0.5) + theme_mike() + 
  labs(x=expression(paste("log"[2], " Fold Change")),
       y=expression(paste("-log"[10], " Adjusted p-value")))+
  geom_text_repel(aes(label=raw.gene.top),
                  colour='black') + 
  ylim(c(0, 150)) + xlim(c(-12, 12)) +
  geom_hline(mapping=aes(yintercept=-log(0.01)), linetype='dashed', colour='purple')

ggsave(raw.volcano,
       filename="../PancreasDE/GSE81076_volcano.png",
       height=4.75, width=4.25, dpi=300)

correct.volcano <- ggplot(correct.de.merge, aes(x=logFC, y=-log10(adj.P.Val))) +
  geom_point(alpha=0.5) + theme_mike() + 
  labs(x=expression(paste("log"[2], " Fold Change")),
       y=expression(paste("-log"[10], " Adjusted p-value"))) +
  geom_text_repel(aes(label=correct.gene.top),
                  colour='black') + 
  ylim(c(0, 350)) + xlim(c(-12, 12)) +
  geom_hline(mapping=aes(yintercept=-log(0.01)), linetype='dashed', colour='purple')

ggsave(correct.volcano,
       filename="../PancreasDE/Corrected_volcano.png",
       height=4.75, width=4.25, dpi=300)

volcano.grid <- plot_grid(raw.volcano, correct.volcano)
volcano.grid
```

What about genes that are only DE in the total data set, and not in each individual one?  The raw DE genes should be a subset of the MNN corrected and combined DE genes.

```{r, echo=FALSE, warning=FALSE, message=FALSE, fig.width=7.75, fig.height=4.75}
raw.de.genes <- raw.de.merge$ensembl_gene_id[raw.de.merge$Sig == 1]
correct.de.genes <- correct.de.merge$ensembl_gene_id[correct.de.merge$Sig == 1]
gse81076.false <- setdiff(raw.de.genes, correct.de.genes)

grid.newpage()
vn.de <- draw.pairwise.venn(area1=length(raw.de.genes), area2=length(correct.de.genes),
                            cross.area=length(intersect(raw.de.genes, correct.de.genes)),
                            category=c("GSE81076 DE\ngenes", "MNN DE\ngenes"), lwd=c(2, 2),
                            col=c("#fdb462", "#a6cee3"), fill=c("#fdb462", "#a6cee3"),
                            alpha=c(0.3, 0.3), cat.pos=c(220, 130), cat.dist=c(0.1, 0.1),
                            cat.cex=c(2, 2), cex=c(1.5, 1.5, 1.5),
                            euler.d=TRUE, scaled=TRUE, margin=(0.1), ext.text=FALSE)
grid.draw(vn.de)

png("../PancreasDE/GSE81076_Venn.png",
    height=7.75, width=7.75, res=300, units="in")
grid.draw(vn.de)
dev.off()
```
Thus we can see that by appropriately combining data sets and correcting for batch effects between them we can construct new cell labels that improves our power to detect differentially expressed genes.  There are 2 genes identified (_GCG_ and _CRYBA2_), which might be indicative of either cell-contamination from $\beta$-islets in GSE81076, or a false-positive result.

Let's look at the overlaps for each individual study with the overall combined DE genes.

```{r, echo=FALSE, warning=FALSE, message=FALSE, fig.width=7.75, fig.height=4.75}
indiv.de.genes <- gene_symbol$ensembl_gene_id[gene_symbol$external_gene_name %in% rownames(gse85241.raw.de.res[gse85241.raw.de.res$Sig == 1, ])]
gse85241.false <- setdiff(indiv.de.genes, correct.de.genes)

grid.newpage()
vn.de.sep <- draw.pairwise.venn(area1=length(indiv.de.genes), area2=length(correct.de.genes),
                            cross.area=length(intersect(indiv.de.genes, correct.de.genes)),
                            category=c("GSE85241 DE\ngenes", "MNN DE\ngenes"), lwd=c(2, 2),
                            col=c("#fdb462", "#a6cee3"), fill=c("#fdb462", "#a6cee3"),
                            alpha=c(0.3, 0.3), cat.pos=c(220, 130), cat.dist=c(0.1, 0.1),
                            cat.cex=c(2, 2), cex=c(1.5, 1.5, 1.5),
                            euler.d=TRUE, scaled=TRUE, margin=(0.1), ext.text=FALSE)
grid.draw(vn.de.sep)

png("../PancreasDE/GSE85241_Venn.png",
    height=7.75, width=7.75, res=300, units="in")
grid.draw(vn.de.sep)
dev.off()
```


```{r, echo=FALSE, warning=FALSE, message=FALSE, fig.width=7.75, fig.height=4.75}
indiv.de.genes <- gene_symbol$ensembl_gene_id[gene_symbol$external_gene_name %in% rownames(gse86473.raw.de.res[gse86473.raw.de.res$Sig == 1, ])]
gse86473.false <- setdiff(indiv.de.genes, correct.de.genes)

grid.newpage()
vn.de.sep <- draw.pairwise.venn(area1=length(indiv.de.genes), area2=length(correct.de.genes),
                            cross.area=length(intersect(indiv.de.genes, correct.de.genes)),
                            category=c("GSE86473 DE\ngenes", "MNN DE\ngenes"), lwd=c(2, 2),
                            col=c("#fdb462", "#a6cee3"), fill=c("#fdb462", "#a6cee3"),
                            alpha=c(0.3, 0.3), cat.pos=c(220, 130), cat.dist=c(0.1, 0.1),
                            cat.cex=c(2, 2), cex=c(1.5, 1.5, 1.5),
                            euler.d=TRUE, scaled=TRUE, margin=(0.1), ext.text=FALSE)
grid.draw(vn.de.sep)

png("../PancreasDE/GSE86473_Venn.png",
    height=7.75, width=7.75, res=300, units="in")
grid.draw(vn.de.sep)
dev.off()
```

For all 3 studies individually there are 2-14 genes that do not appear in the DE set based on the batch corrected cluster labels.  None of these genes overlap with each other, potentially giving a rough idea of the false positive rate within each individual study DE analysis.

```{r, echo=FALSE, warning=FALSE, message=FALSE}
# write out a list of the corrected DE genes
write.table(correct.de.genes,
            file="../PancreasDE/Corrected_DE_genes.tsv",
            sep="\t", quote=F, row.names=F)
```

